File: bible-tools/src/main/java/rawcomposition/bibletools/info/BibleToolsApp.java
package rawcomposition.bibletools.info;

import android.app.Application;
import android.content.Context;

import io.realm.Realm;
import io.realm.RealmConfiguration;
import io.realm.RealmMigration;
import rawcomposition.bibletools.info.util.FontCache;
import rawcomposition.bibletools.info.util.enums.FontWeight;

/**
 * Created by tinashe on 2015/07/26.
 */
public class BibleToolsApp extends Application {

    private static final String TAG = BibleToolsApp.class.getName();

    private static final int REALM_VERSION = 2;

    private static Context context;

    @Override
    public void onCreate() {
        super.onCreate();

        context = getApplicationContext();

        RealmConfiguration config = new RealmConfiguration.Builder(this)
                .migration(new BibleToolsRealmMigration())
                .schemaVersion(REALM_VERSION)
                .build();
        Realm.setDefaultConfiguration(config);

        //Init Fonts
        for (FontWeight weight : FontWeight.values()) {
            FontCache.getInstance().addFont(weight.getName(), weight.getFileName());
        }
    }

    class BibleToolsRealmMigration implements RealmMigration {

        @Override
        public long execute(Realm realm, long version) {
            return REALM_VERSION;
        }
    }

    public static Context getAppContext() {
        return BibleToolsApp.context;
    }
}

File: bible-tools/src/main/java/rawcomposition/bibletools/info/BibleToolsApp.kt
package rawcomposition.bibletools.info

import android.app.Activity
import android.app.Application
import android.support.v4.app.Fragment
import com.crashlytics.android.Crashlytics
import com.crashlytics.android.answers.Answers
import com.crashlytics.android.core.CrashlyticsCore
import dagger.android.AndroidInjector
import dagger.android.DispatchingAndroidInjector
import dagger.android.HasActivityInjector
import dagger.android.support.HasSupportFragmentInjector
import io.fabric.sdk.android.Fabric
import rawcomposition.bibletools.info.di.DaggerBibleToolsAppComponent
import timber.log.Timber
import javax.inject.Inject

class BibleToolsApp : Application(), HasActivityInjector, HasSupportFragmentInjector {

    @Inject
    lateinit var activityInjector: DispatchingAndroidInjector<Activity>

    @Inject
    lateinit var fragmentInjector: DispatchingAndroidInjector<Fragment>

    override fun activityInjector(): AndroidInjector<Activity> = activityInjector

    override fun supportFragmentInjector(): AndroidInjector<Fragment> = fragmentInjector


    override fun onCreate() {
        super.onCreate()

        AppInjector.init(this)
    }

    object AppInjector {

        fun init(app: BibleToolsApp) {

            val core = CrashlyticsCore.Builder().disabled(BuildConfig.DEBUG).build()
            Fabric.with(app, Crashlytics.Builder().core(core).build(), Answers())

            if (BuildConfig.DEBUG) {
                Timber.plant(Timber.DebugTree())
            }

            DaggerBibleToolsAppComponent.builder()
                    .application(app)
                    .build()
                    .inject(app)
        }
    }
}
